activerecord/lib/active_record/associations/builder/has_and_belongs_to_many.rb

```diff
@@ -1,13 +1,18 @@
 # frozen_string_literal: true
 
 module ActiveRecord::Associations::Builder # :nodoc:
-  class HasAndBelongsToMany # :nodoc:
+  class HasAndBelongsToMany < CollectionAssociation # :nodoc:
     attr_reader :lhs_model, :association_name, :options
 
     def initialize(association_name, lhs_model, options)
       @association_name = association_name
       @lhs_model = lhs_model
       @options = options
+      validate_options
+    end
+
+    def self.valid_options(options)
+      super + %i[join_table association_foreign_key] - %i[primary_key query_constraints inverse_of dependent]
```

Comment:
Class methods should be all defined before any of the instance methods. Can you please move?